benchmark.py

```diff
@@ -508,30 +672,36 @@ def main(phase, host, port, user, password, database, data_dir, query_root, dbge
         if create_schema(query_root, host, port, database, user, password):
             print("could not create schema.")
             exit(1)
-        result.setCreateTime(result.stopTimer())
+        result.setMetric("create_schema: ", result.stopTimer())
         print("done creating schemas")
         result.startTimer()
         if load_tables(query_root, data_dir, host, port, database, user, password):
             print("could not load data to tables")
             exit(1)
-        result.setLoadTime(result.stopTimer())
+        result.setMetric("load_data", result.stopTimer())
         print("done loading data to tables")
         result.startTimer()
         if index_tables(query_root, data_dir, host, port, database, user, password):
             print("could not create indexes for tables")
             exit(1)
-        result.setIndexingTime(result.stopTimer())
+        result.setMetric("index_tables", result.stopTimer())
         print("done creating indexes and foreign keys")
-        result.printLoadTime()
+        result.printMetrics()
     elif phase == "query":
-        # TODO: this thing should run twice and average the results
-        result = Result()
-        if run_power_test(query_root, data_dir, result, host, port, database, user, password):
-            print("Running power test failed")
-            exit(1)
-        print("Power test finished.")
-        result.printSeqTime()
-        # TODO: add throughput test here
+        num_runs = 2 # as per spec, the test should run twice, with a reboot between them
+        for run in range(num_runs):
+            # Power test
+            if run_power_test(query_root, data_dir, host, port, database, user, password,
+                              run, num_streams, verbose, read_only):
+                print("running power test failed")
+                exit(1)
+            # Throughput test
+            if run_throughput_test(query_root, data_dir, host, port, database, user, password,
+                                   run, num_streams, verbose, read_only):
+                print("running throughput test failed")
+                exit(1)
+            if run < num_runs - 1:
+                reboot() # no need to reboot at the last run
```

Comment:
I don't think this fits the specs. Reboot will not help here. The two runs should be identical. Even the data used for refresh streams. I believe we need to call `prepare` again. That is we need to drop the tables, load data and create indexes again. Another solution would be to truncate all tables and load the data. In this case, you wouldn't need to create the indexes as the tables are not dropped. But the `prepare` thing is already implemented. So it would be much easier I believe. The only difference from the prepare phase would be that this doesn't need to be timed. Also, we cannot rely on the fact that the database is local (in fact for our purpose, it would be some remote database managed through Brocolli). So reboot cannot work the way it is implemented right now. A local database is alright only for the testing or for Travis. The code itself should not rely on this.

Reply:
Good point. Will add it as a TODO. For now, I will just remove it, I guess.

Reply:
Though, I just found following in the spec:
> Comment: It is explicitly not permitted to rebuild or reload the test database during the performance test (see Clause 5.1.1.3)

So rebuilding the DB is not an option, if I understand correctly.
How about removing the phases from arg list and running all in one go?
We have anyway the issue, that after running refresh functions the query phase can not be repeated (that is why I added read_only arg to make development easier and not to be forced to generate and load the data again)

Reply:
During it. Yes. The point they are trying to make is that between the power and throughput there should be no reloads. Which is why we use different datasets for them.

Reply:
I think the easiest way right now is that the user runs it as follows:
```
benchmark.py prepare
benchmark.py load
benchmark.py query #note times
benchmark.py load # second time
benchmark.py query #note times for iteration 2
```
The problem with this is how do you organise the results such that it is easy to comprehend.

Reply:
If you check my PR, the code is quite flexible now on how the results folder and the JSON files inside it are organised.

Reply:
Another option would be to add another phase ... metrics.

Reply:
How about merging this and metrics branch to master? Then we could open some issues to address these points.

Reply:
I am sorry I don't understand. How would another phase help? I don't think it makes sense to merge this with the `reboot` function. That is simply incorrect. What we can do is go back to single run for the `query` phase then tackle the second run through another issue or leave it to the user whatever is convinient. I will look at the metrics thing.
(Since its only a single file it might be wise to work on one issue at a time otherwise there could be lots of conflicts.)

Reply:
I thought anothe rpahse for metrics would help to keep things split apart. Running the perf test in query phase and calculating metrics in metrics phase. Otherwise use might get confused if the metrics are related to the last run only or to all runs. But when I think about it today, the extra phase would not solve that issue either.